core/src/main/java/net/opentsdb/query/plan/DefaultQueryPlanner.java

```diff
@@ -807,7 +851,98 @@ public String getMetricForDataSource(final QueryNodeConfig node,
     
     return null;
   }
-  
+
+  @Override
+  public TimeAdjustments getAdjustments(
+          final TimeSeriesDataSourceConfig config) {
+    return recursiveAdjustments(config, null);
+  }
+
+  /**
+   * Helper to traverse the graph and find the proper adjustments if any.
+   * @param config The non-null config to walk back from.
+   * @param adjustments An optional adjustment. Can be null.
+   * @return Adjustments if found, null if not.
+   */
+  protected TimeAdjustments recursiveAdjustments(final QueryNodeConfig config,
+                                                 TimeAdjustments adjustments) {
+    final Set<QueryNodeConfig> predecessors = config_graph.predecessors(config);
+    for (final QueryNodeConfig predecessor : predecessors) {
+      if (predecessor instanceof DownsampleConfig) {
+        if (adjustments == null) {
+          adjustments = new TimeAdjustments();
+          adjustments.downsampleInterval =
+                  ((DownsampleConfig) predecessor).getInterval();
+          if (adjustments.downsampleInterval.equals(DownsampleConfig.AUTO)) {
+            final long deltaSeconds = context.query().endTime().epoch() -
+                    context.query().startTime().epoch();
+            adjustments.downsampleInterval =
+                    DownsampleFactory.getAutoInterval(context().tsdb(),
+                            deltaSeconds,
+                            ((DownsampleConfig) predecessor).getMinInterval());
+          } else if (((DownsampleConfig) predecessor).getRunAll()) {
+            adjustments.downsampleInterval = null;
```

Comment:
Does this mean there are multiple downsample intervals ? If not, we should consider the case where there are multiple downsample intervals.

Reply:
Yeah, we only take the first one (predecessor) as that is what will affect the overlap in the data sources.